### PR TITLE
Add session-manager module

### LINKED ISSUE #1190: module-session-manager
https://github.com/AcalaNetwork/Acala/blob/32c50c18f1040ed9876952f5dcaa745e8d4926ea/runtime/karura/src/lib.rs#L268

Karura is currently having 6h session time which is too long. We want it to be 1h initially and maybe in future change to 3h or longer.

We can't just change it because it will mess-up with session index. So need a new implementation to handle the change of session time.

https://github.com/AcalaNetwork/Acala/blob/32c50c18f1040ed9876952f5dcaa745e8d4926ea/runtime/karura/src/lib.rs#L277-L278

- module-session-manager
  - Config
  - Storages
    - SessionDuration: BlockNumber
    - SessionIndex: SessionIndex
    - SessionDurationChanges: Vec\<(BlockNumber, SessionIndex, BlockNumber)>
      - (the block number that the new change become effective, the session number, new session duration)
  - Call
    - fn schedule_session_duration(origin, start_session: SessionIndex, duration: BlockNumber)
      - ensure root
      - ensure session index is not in past
  - Pallet
    - fn on_initialize
      - apply session change if needed
    - impl ShouldEndSession
    - impl EstimateNextSessionRotation

Need integration tests
Need migration to ensure it is compatible with what Karura is currently have

### PR DESCRIPTION
**Pull request by zjb0807**
Closes #1190

### CODE REVIEW

File: modules/session-manager/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+		let period = Self::session_duration();
+
+		// NOTE: we add one since we assume that the current block has already elapsed,
+		// i.e. when evaluating the last block in the session the progress should be 100%
+		// (0% is never returned).
+		let progress = if now >= offset {
+			let current = (now - offset) % period + One::one();
+			Some(Permill::from_rational(current, period))
+		} else {
+			Some(Permill::from_rational(now + One::one(), offset))
```
Comment (xlc) [inline modules/session-manager/src/lib.rs:204]: I think it is ok to just return `None` in this case

File: modules/session-manager/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+		};
+
+		// Weight note: `estimate_current_session_progress` has no storage reads and trivial
```
Comment (xlc) [inline modules/session-manager/src/lib.rs:207]: offset and period are storage read?

File: modules/session-manager/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+			}
+		} else {
+			offset
+		};
+
+		// Weight note: `estimate_next_session_rotation` has no storage reads and trivial
+		// computational overhead. There should be no risk to the chain having this weight value be
+		// zero for now. However, this value of zero was not properly calculated, and so it would be
+		// reasonable to come back here and properly calculate the weight of this function.
+		(Some(next_session), Zero::zero())
```
Comment (xlc) [inline modules/session-manager/src/lib.rs:237]: two reads

**Review by xlc — COMMENTED** [review]
Need runtime migration code to initialize SessionDuration & DurationOffset for Karura

**Comment by zjb0807** [comment]
Tested it with the launch.
But the `try-runtime` command does not work.
```
cargo run --features with-karura-runtime --features try-runtime -- try-runtime --chain=karura --wasm-execution=compiled on-runtime-upgrade live "wss://karura.api.onfinality.io/public-ws" -s snapshot.bin
```
return
```
2021-07-12 20:48:49 initializing remote client to "wss://karura.api.onfinality.io/public-ws"
2021-07-12 20:48:50 scraping key-pairs from remote @ 0x8650efc526589f063e62200925ba9a2e6940c16eaa8937dd866ff0b04bdf87c0
2021-07-12 20:48:50 downloading data for all modules.
2021-07-12 20:50:03 Querying a total of 85580 keys
2021-07-12 20:51:28 writing to state snapshot file "snapshot.bin"
2021-07-12 20:51:28 extending externalities with 1 manually injected keys
2021-07-12 20:51:28 injecting a total of 85581 keys
Error: Input("failed to execute \'TryRuntime_on_runtime_upgrade\' due to Other(\"Exported method TryRuntime_on_runtime_upgrade is not found\")")
```

**Comment by zjb0807** [comment]
use `mandala-runtime` to test migration is the same as `karura-runtime` because there is no CustomOnRuntimeUpgrade in runtime.

```
cargo run --features with-mandala-runtime --features try-runtime -- try-runtime --wasm-execution=compiled on-runtime-upgrade live "wss://karura.api.onfinality.io/public-ws" -s snapshot.bin
```


```
Running `target/debug/acala try-runtime --wasm-execution=compiled on-runtime-upgrade live 'wss://karura.api.onfinality.io/public-ws' -s snapshot.bin`
2021-07-13 16:27:45 initializing remote client to "wss://karura.api.onfinality.io/public-ws"
2021-07-13 16:27:46 scraping key-pairs from remote @ 0x25595bbf83e48732913124b387b6f61b799e90c563cb8d0dae60eef48d91112d
2021-07-13 16:27:46 downloading data for all modules.
2021-07-13 16:29:12 Querying a total of 85722 keys
2021-07-13 16:30:46 writing to state snapshot file "snapshot.bin"
2021-07-13 16:30:46 extending externalities with 1 manually injected keys
2021-07-13 16:30:46 injecting a total of 85723 keys
2021-07-13 16:31:17 ⚠️ System declares internal migrations (which *might* execute), setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for Sudo, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ⚠️ SessionManager declares internal migrations (which *might* execute), setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 Migrating session-manager v1
2021-07-13 16:31:17 Completed session-manager migration to v1
2021-07-13 16:31:17 ✅ no migration for AuraExt, setting storage version to PalletVersion { major: 0, minor: 1, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for Aura, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for Session, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for CollatorSelection, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for Authorship, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for EvmManager, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for EvmAccounts, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for EVMBridge, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for EVM, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for UnknownTokens, setting storage version to PalletVersion { major: 0, minor: 4, patch: 1 }
2021-07-13 16:31:17 ✅ no migration for XTokens, setting storage version to PalletVersion { major: 0, minor: 4, patch: 1 }
2021-07-13 16:31:17 ✅ no migration for DmpQueue, setting storage version to PalletVersion { major: 0, minor: 1, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for CumulusXcm, setting storage version to PalletVersion { major: 0, minor: 1, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for PolkadotXcm, setting storage version to PalletVersion { major: 0, minor: 1, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for XcmpQueue, setting storage version to PalletVersion { major: 0, minor: 1, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for ParachainInfo, setting storage version to PalletVersion { major: 0, minor: 1, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for ParachainSystem, setting storage version to PalletVersion { major: 0, minor: 1, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for CompoundCash, setting storage version to PalletVersion { major: 1, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for Starport, setting storage version to PalletVersion { major: 1, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for ChainSafeTransfer, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for ChainBridge, setting storage version to PalletVersion { major: 0, minor: 0, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for RenVmBridge, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for AirDrop, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for NFT, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for Incentives, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for HomaValidatorListModule, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for PolkadotBridge, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for StakingPool, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for NomineesElection, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for Homa, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for EmergencyShutdown, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for CdpEngine, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for CdpTreasury, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for Honzon, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for Loans, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for AuctionManager, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for Dex, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for Prices, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for OrmlNFT, setting storage version to PalletVersion { major: 0, minor: 4, patch: 1 }
2021-07-13 16:31:17 ✅ no migration for Rewards, setting storage version to PalletVersion { major: 0, minor: 4, patch: 1 }
2021-07-13 16:31:17 ✅ no migration for Auction, setting storage version to PalletVersion { major: 0, minor: 4, patch: 1 }
2021-07-13 16:31:17 ✅ no migration for OperatorMembershipBand, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for BandOracle, setting storage version to PalletVersion { major: 0, minor: 4, patch: 1 }
2021-07-13 16:31:17 ✅ no migration for OperatorMembershipAcala, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for AcalaOracle, setting storage version to PalletVersion { major: 0, minor: 4, patch: 1 }
2021-07-13 16:31:17 ✅ no migration for Democracy, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for ElectionsPhragmen, setting storage version to PalletVersion { major: 4, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for Authority, setting storage version to PalletVersion { major: 0, minor: 4, patch: 1 }
2021-07-13 16:31:17 ✅ no migration for TechnicalCommitteeMembership, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for TechnicalCommittee, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for HomaCouncilMembership, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for HomaCouncil, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for FinancialCouncilMembership, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for FinancialCouncil, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for GeneralCouncilMembership, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for GeneralCouncil, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for GraduallyUpdate, setting storage version to PalletVersion { major: 0, minor: 4, patch: 1 }
2021-07-13 16:31:17 ✅ no migration for Indices, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for Proxy, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for Recovery, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for Multisig, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for Utility, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for Tips, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for Bounties, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for Treasury, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for TransactionPayment, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for Vesting, setting storage version to PalletVersion { major: 0, minor: 4, patch: 1 }
2021-07-13 16:31:17 ✅ no migration for Currencies, setting storage version to PalletVersion { major: 1, minor: 1, patch: 2 }
2021-07-13 16:31:17 ✅ no migration for Tokens, setting storage version to PalletVersion { major: 0, minor: 4, patch: 1 }
2021-07-13 16:31:17 ✅ no migration for Balances, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for Scheduler, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 ✅ no migration for Timestamp, setting storage version to PalletVersion { major: 3, minor: 0, patch: 0 }
2021-07-13 16:31:17 TryRuntime_on_runtime_upgrade executed without errors. Consumed weight = 1952000000000, total weight = 128000000000000 (0.01525)

**Comment by xlc** [comment]
will merge this after the next runtime release is out

**Review by shaunxw — COMMENTED** [review]
We need a unit test to catch `EstimateNextSessionFailed` error.

**Review by shaunxw — APPROVED** [review]

**Pull request merged**